utils/utils.c

```diff
@@ -154,10 +155,14 @@ int remove_directory(char *dirname)
 			continue;
 
 		snprintf(buf, sizeof(buf), "%s/%s", dirname, ent->d_name);
-		if (unlink(buf) < 0) {
-			saved_errno = errno;
-			ret = -1;
-			break;
+		if (!stat(buf, &statbuf)) {
+			if (S_ISDIR(statbuf.st_mode))
+				ret = remove_directory(buf);
```

Comment:
What if `stat` or (recursive) `remove_directory` failed?

Reply:
Ah, I missed it.. what about the below change ?
(Because of duplication, I used `failed` )
```patch
diff --git a/utils/utils.c b/utils/utils.c
index cb08634..b51b31a 100644
--- a/utils/utils.c
+++ b/utils/utils.c
@@ -156,14 +156,19 @@ int remove_directory(char *dirname)
 
                 snprintf(buf, sizeof(buf), "%s/%s", dirname, ent->d_name);
                 if (!stat(buf, &statbuf)) {
-                        if (S_ISDIR(statbuf.st_mode))
-                                ret = remove_directory(buf);
+                        if (S_ISDIR(statbuf.st_mode) &&
+                            remove_directory(buf) < 0)
+                                goto failed;
                         else if (unlink(buf) < 0) {
                                 saved_errno = errno;
-                                ret = -1;
-                                break;
+                                goto failed;
                         }
                 }
+                else {
+failed:
+                        ret = -1;
+                        break;
+                }
         }
 
         closedir(dp);

```

Reply:
Looks ok, but you also need to save errno for all error cases.

Reply:
@namhyung , I modified this PR 😄 
I'd appreciate it, if you check it.

Reply:
In addition, I found my mistake that is twice elimination tries for same directory.

In the case "it is directory && successfully remove it",
the statement `else if (unlink(buf) < 0)` can run as well.

For example, even though 'uftrace.data.old/test' was already removed
by remove_directory(), unlink() can try to remove it, again.

So we can see the warnning `No such file or directory` like below:
```
  $ mkdir -p uftrace.data.old/test

  $ uftrace record hello
  WARN: removing old directory failed: No such file or directory
```

So, I fixed it as well in new PR.